crates/ruff/src/args.rs

```diff
@@ -491,97 +507,184 @@ impl From<&LogLevelArgs> for LogLevel {
+    config_overrides: Configuration,
+    /// Overrides provided via dedicated flags such as --line-length etc.
+    per_flag_overrides: ExplicitConfigOverrides,
+}
+
+impl ConfigArgs {
+    pub fn config_file(&self) -> Option<&PathBuf> {
+        self.config_file.as_ref()
+    }
+
+    fn from_cli_options(
+        config_options: Vec<ConfigOption>,
+        per_flag_overrides: ExplicitConfigOverrides,
+        isolated: bool,
+    ) -> Result<Self, anyhow::Error> {
+        let mut new = Self {
+            per_flag_overrides,
+            ..Self::default()
+        };
+
+        for option in config_options {
+            match option {
+                ConfigOption::ConfigOverride(overridden_option) => {
+                    let overridden_option = Arc::try_unwrap(overridden_option)
+                        .unwrap_or_else(|option| Options::clone(&option));
+                    new.config_overrides = new.config_overrides.combine(
+                        Configuration::from_options(overridden_option, &path_dedot::CWD)?,
+                    );
+                }
+                ConfigOption::PathToConfigFile(path) => {
+                    if isolated {
+                        let error = anyhow!(
+                            "Cannot specify `--isolated` and also specify a configuration file"
+                        );
+                        let context = format!(
+                            "Both `--isolated` and `--config={}` were specified on the command line",
+                            path.display()
+                        );
+                        return Err(error.context(context));
+                    }
+                    if let Some(ref config_file) = new.config_file {
+                        let (first, second) = (config_file.display(), path.display());
+                        let mut error = clap::Error::new(clap::error::ErrorKind::ValueValidation);
+                        error.insert(
+                            clap::error::ContextKind::InvalidArg,
+                            clap::error::ContextValue::String("--config".to_string()),
+                        );
+                        error.insert(
+                            clap::error::ContextKind::InvalidValue,
+                            clap::error::ContextValue::String(second.to_string()),
+                        );
+                        let tips = vec![
+                            "Cannot specify more than one configuration file on the command line"
+                                .into(),
+                            format!("Remove either `--config={first}` or `--config={second}`")
+                                .into(),
+                        ];
+                        error.insert(
+                            clap::error::ContextKind::Suggested,
+                            clap::error::ContextValue::StyledStrs(tips),
+                        );
+                        return Err(error.into());
```

Comment:
Using `clap` here means we get nicely formatted multiline error messages with the `tip: ` bit at the beginning:

![image](https://github.com/astral-sh/ruff/assets/66076021/9c080a05-e872-4bd0-b34b-8dbd680f33d5)

Unfortunately, because we're using `clap` _after_ the initial argument parsing here, we don't get the nice colours that we do elsewhere -- this is what it looks like when you use `clap` in the initial argument parsing:

![image](https://github.com/astral-sh/ruff/assets/66076021/12e2d6b3-fbd1-498a-b322-49d8998546c4)

I can't really see an easy solution to that

Reply:
Hmm, I have an idea. Will try it out tomorrow...

Reply:
On second though, I think I will leave this, too, to a subsequent PR.